File: primer1/primer1.js
const temperatureConversion = (temperature, fromScale, toScale) => {
    // Explicit validation for null, undefined, and non-numeric inputs
    if (temperature === null || temperature === undefined ||
        (typeof temperature === 'string' && isNaN(Number(temperature)))) {
        throw new Error('Invalid temperature input');
    }

    // Convert input to number
    const numTemp = Number(temperature);

    // Normalise conversion type to uppercase
    toScale = String(toScale).toUpperCase();
    fromScale = String(fromScale).toUpperCase();

    // Validate conversion type
    if (!['C', 'F', 'K'].includes(toScale) || !['C', 'F', 'K'].includes(fromScale)) {
        throw new Error('Invalid conversion type or input scale');
    }

    // Conversion helpers
    const toFahrenheit = (celsius) => (celsius * 9 / 5) + 32;
    const toCelsius = (fahrenheit) => (fahrenheit - 32) * 5 / 9;
    const toKelvin = (celsius) => celsius + 273.15;

    // Conversion logic
    switch (toScale) {
        case 'C':
            return fromScale === 'F' ? toCelsius(numTemp) : fromScale === 'K' ? numTemp - 273.15 : numTemp;
        case 'F':
            return fromScale === 'C' ? toFahrenheit(numTemp) : fromScale === 'K' ? toFahrenheit(numTemp - 273.15) : numTemp;
        case 'K':
            // Handles both Celsius and Fahrenheit to Kelvin
            return fromScale === 'K' ? numTemp : toKelvin(fromScale === 'F' ? toCelsius(numTemp) : numTemp);
    }
};


console.log(temperatureConversion(60.1, 'C', 'F'))
console.log(temperatureConversion(60.1, 'C', 'C'))
console.log(temperatureConversion(60.1, 'F', 'C'))
console.log(temperatureConversion(-17.78, 'C', 'F'))

export default temperatureConversion;

File: primer1/primer1.py
def temperature_conversion(temperature, from_scale, to_scale):
    """
    Converts temperature between Celsius, Fahrenheit, and Kelvin.

    Args:
      temperature: The temperature value to convert.
      from_scale: The scale of the input temperature ('C', 'F', or 'K').
      to_scale: The scale to convert the temperature to ('C', 'F', or 'K').

    Returns:
      The converted temperature, or None if invalid input is provided.
    """
    try:
        temperature = float(temperature)  # Ensure temperature is a number
        from_scale = from_scale.upper()
        to_scale = to_scale.upper()

        # Validate scales before proceeding
        valid_scales = {'C', 'F', 'K'}
        if from_scale not in valid_scales or to_scale not in valid_scales:
            raise ValueError("Invalid conversion type")

        if from_scale == to_scale:
            return temperature

        if from_scale == 'C':
            if to_scale == 'F':
                return (temperature * 9 / 5) + 32
            elif to_scale == 'K':
                return temperature + 273.15
        elif from_scale == 'F':
            if to_scale == 'C':
                return (temperature - 32) * 5 / 9
            elif to_scale == 'K':
                return (temperature - 32) * 5 / 9 + 273.15
        elif from_scale == 'K':
            if to_scale == 'C':
                return temperature - 273.15
            elif to_scale == 'F':
                return (temperature - 273.15) * 9 / 5 + 32

    except ValueError as e:
        raise ValueError(str(e))
    except TypeError:
        raise ValueError("Invalid temperature input")